File: src/leetcode/MergekSortedLists.java
/**
 * Problem link: https://leetcode.com/problems/merge-k-sorted-lists
 *
 * Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
 */

import java.util.*;
public class MergekSortedLists {
    public static void main(String[] args) {
    }

    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode head = null;
        MinHeap<ListNode> minHeap = new MinHeap<>(1024, new ListNodeComparator());
        int length = lists.length;
        for(int i = 0 ; i < length ; ++ i) {
            if(lists[i] != null) {
                minHeap.insert(lists[i]);
            }
        }

        ListNode result = null;

        while(!minHeap.isEmpty()) {
            ListNode top = minHeap.peek();
            top = minHeap.pop(top.next);
            if(result == null) {
                result = head = top;
            } else {
                head.next = top;
                head = head.next;
            }
        }

        return result;
    }

    class ListNodeComparator implements Comparator<ListNode> {
        public int compare(ListNode o1, ListNode o2) {
            if(o1.val < o2.val) {
                return -1;
            } else if (o1.val == o2.val) {
                return 0;
            } else {
                return 1;
            }
        }
    }

    class MinHeap<T> {
        private Object[] array;
        private int capacity;
        private int size;
        private Comparator<T> comp;

        public MinHeap(int capacity, Comparator<T> comp) {
            array = new Object[capacity];
            this.capacity = capacity;
            this.size = 0;
            this.comp = comp;
        }

        public boolean isEmpty() {
            return size == 0;
        }

        public void insert(T element) {
            ensure(size + 1);
            array[size] = element;
            siftUp(size);
            ++ size;
        }

        public T peek() {
            return (T)array[0];
        }

        public T pop() {
            T result = (T)array[0];
            array[0] = array[size - 1];
            -- size;
            siftDown(0);
            return result;
        }

        public T pop(T replace) {
            if(replace == null) {
                return pop();
            } else {
                T result = (T)array[0];
                array[0] = replace;
                siftDown(0);
                return result;
            }
        }

        void ensure(int size) {
            while(size >= capacity) {
                Object[] newArray = new Object[capacity + 128];
                System.arraycopy(array, 0, newArray, 0, array.length);
                capacity += 128;
                array = newArray;
            }
        }

        void siftUp(int pos) {
            int child = pos;
            int parent = (child - 1) >>  1;
            while(child > 0) {
                if(comp.compare((T)array[parent], (T)array[child]) > 0) {
                    swap(parent, child);
                } else {
                    return;
                }
                child = parent;
                parent = (child - 1) >> 1;
            }
        }

        void siftDown(int pos) {
            int parent = pos;
            int child = (parent << 1) + 1;

            while(child < size) {
                if(child < size - 1) {
                    child = comp.compare((T)array[child], (T)array[child + 1]) < 0 ? child : child + 1;
                }

                if(comp.compare((T)array[parent], (T)array[child]) > 0) {
                    swap(parent, child);
                } else {
                    return;
                }

                parent = child;
                child = (parent << 1) + 1;
            }
        }

        void swap(int pos1, int pos2) {
            Object tmp = array[pos1];
            array[pos1] = array[pos2];
            array[pos2] = tmp;
        }
    }
}

File: src/leetcode/merge_k_sorted_lists.cpp
/**
 * Problem link: https://leetcode.com/problems/merge-k-sorted-lists
 *
 * Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
 */

#include <stdio.h>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
    public:
        ListNode* mergeKLists(vector<ListNode*>& lists) {
            if(lists.size() == 0) {
                return NULL;
            }
            init(lists);
            ListNode* result = pop();
            ListNode* iter = result;
            while(iter != NULL) {
                ListNode* node = pop();
                iter->next = node;
                iter = iter->next;
            }

            return result;
        }
    private:
        int* array;
        int array_size;
        vector<ListNode*> iterator;

        void init(vector<ListNode*>& lists) {
            int size = lists.size();
            array = new int[size];
            array_size = size;
            iterator = lists;

            int maxId = 0;
            for(int i = 1 ; i < size ; ++ i) {
                maxId = compare(iterator, maxId, i) < 0 ? maxId : i;
            }

            for(int i = 0 ; i < array_size ; ++ i) {
                array[i] = maxId;
            }

            for(int i = array_size - 1 ; i >= 0 ; -- i) {
                adjust(i);
            }
        }

        ListNode* pop() {
            ListNode* result = getNode(array[0]);
            if(result == NULL) {
                return result;
            }
            adjust(array[0]);
            return result;
        }

        void adjust(int index) {
            int t = (index + array_size) >> 1;
            while(t > 0) {
                if(compare(iterator, array[t], index) < 0) {
                    int tmp = index;
                    index = array[t];
                    array[t] = tmp;
                }

                t = t >> 1;
            }

            array[0] = index;
        }

        int compare(vector<ListNode*>& lists, int index1, int index2) {
            ListNode *node1 = lists[index1];
            ListNode *node2 = lists[index2];
            if(node1 == NULL && node2 == NULL) {
                return index1 < index2 ? -1 : 1;
            } else if (node1 == NULL && node2 != NULL) {
                return 1;
            } else if (node1 != NULL && node2 == NULL) {
                return -1;
            } else {
                int value1 = node1->val;
                int value2 = node2->val;
                if(value1 < value2) {
                    return -1;
                } else if(value1 > value2) {
                    return 1;
                } else {
                    return index1 < index2 ? -1 : 1;
                }
            }
        }

        ListNode* getNode(int index) {
            ListNode *result = iterator[index];
            if(result == NULL) {
                return NULL;
            } else {
                iterator[index] = result->next;
                return result;
            }
        }
};
